### DISCUSSION TITLE
pytest can't seem to access Redis session store

### DISCUSSION THREAD

**Discussion by behai-nguyen** [alex-oleshkevich/starsessions#72] [Q&A] [answered]
Hi,

Thank you for this library. I'm learning FastAPI and I find this library very helpful.

I have used its Redis session store without any problem in my application proper. However, my integration test using ``pytest`` does not seem to be able to "access" the Redis session store.

What have I missed please?

Content of ``/fastapi_learning/tests/conftest.py``:

```python
import pytest

from fastapi.testclient import TestClient

from . import test_main

@pytest.fixture(scope='module')
def app():
    return test_main.app

@pytest.fixture(scope='module')
def test_client(app):
    return TestClient(app)
```

Following is the ``/auth/logout`` endpoint which I am writing test for:

```python
def __login_page_context(inc_not_login_msg=False):
    context = {"title": "FastAPI Login"}
    if inc_not_login_msg:
        context.update({"message": "This session has not been logged in before"})

    return context

@router.post("/logout", response_class=HTMLResponse)
async def logout(request: Request):
    context = __login_page_context(
        inc_not_login_msg=request.session.get("access_token")==None)
    
    request.session.clear()
    
    return templates.TemplateResponse(request=request, name="auth/login.html", context=context)
```

When the test gets to this endpoint, ``request.session`` is ``{}``.

The integration test is not able to clear the session content from Redis. While the application is able to. That is, the application works as expected.

The call below has no effect during test, the corresponding Redis session does not get cleared:

```python
    request.session.clear()
```	

And this is my test method:

```python
@pytest.mark.auth_integration
def test_integration_valid_login(test_client):
    """
    Test /auth/token path with a valid credential.
    """

    try:
        login_data = {
            "username": "behai_nguyen@hotmail.com",
            "password": "password"
        }
        response = test_client.post(  
            "/auth/token", 
            data=login_data,
            headers={ 'Content-Type': 'application/x-www-form-urlencoded'}
        )

        assert response != None
        assert response.status_code == HTTPStatus.OK.value

        status = response.json()
        assert status['access_token'] == 'behai_nguyen@hotmail.com'
        assert status['token_type'] == 'bearer'

        print("\ncookie session -> ", test_client.cookies.get('session'))

    finally:
        # Logout. Clean up server sessions.
        test_client.post('/auth/logout')
```

``test_client.cookies.get('session')`` does have a valid session Id, which matches
the one in Redis, for example ``e87901248334d8f8e8d2b3fb4175b8c4``.

And ``test_client.post('/auth/logout')`` does go to the endpoint code.

Thank you and best regards,

...behai.

**Comment by behai-nguyen** [comment, marked as answer]
With the helps of ChatGPT, I have the solution for it:

```python
@pytest.mark.behai_only
def test_integration_valid_login(test_client):
    """
    Test /auth/token path with a valid credential.
    """

    try:
        login_data = {
            'username': 'behai_nguyen@hotmail.com',
            'password': 'password'
        }
        response = test_client.post(  
            '/auth/token', 
            data=login_data,
            headers={'Content-Type': 'application/x-www-form-urlencoded'}
        )

        assert response != None
        assert response.status_code == HTTPStatus.OK.value

        status = response.json()
        assert status['access_token'] == 'behai_nguyen@hotmail.com'
        assert status['token_type'] == 'bearer'

        session_cookie = response.cookies.get('session')

    finally:
        test_client.cookies = {'session': session_cookie}
        test_client.post('/auth/logout')
```

**Comment by alex-oleshkevich** [comment]
I don't know how your application is configured, but check if you have `cookie_https_only` set to False in tests. In the default configuration, the cookie works only in secure HTTPS, unit tests typically use HTTP, so the cookie is just not available there.

Refer to read me for details.

**Reply by behai-nguyen** [reply]
Hi @alex-oleshkevich, 

Thank you for your reply. I did do some debuggings. I understand the code a bit better.

But ChatGPT has helped me solved this. Basically, we need to set ``test_client.cookies``:

```python
session_cookie = response.cookies.get('session')

test_client.cookies = {'session': session_cookie}
test_client.post('/auth/logout')
```
Then in the session middleware, the dunder ``__call__``  method is able to correctly identify the session Id. 

I did set cookies during my debugs, but I created cookies and passed it in. I resulted in runtime error.

Thank you for such great library.

Best regards,

...behai.
